File: solutions/javascript/clock/1/clock.js
export class Clock {
    #hours;
    #minutes;

    /** @param {number} hours */
    constructor(hours, minutes = 0) {
        this.#hours = hours;
        this.#minutes = minutes;
        this.#normalize();
    }

    #normalize() {
        let totalMinutes = this.#hours * 60 + this.#minutes;
        totalMinutes = ((totalMinutes % (24 * 60)) + 24 * 60) % (24 * 60);
        this.#hours = Math.floor(totalMinutes / 60);
        this.#minutes = totalMinutes % 60;
    }

    toString() {
        /** @param {number} num */
        const pad = (num) => num.toString().padStart(2, '0');
        return `${pad(this.#hours)}:${pad(this.#minutes)}`;
    }

    /** @param {number} minutes */
    plus(minutes) {
        return new Clock(this.#hours, this.#minutes + minutes);
    }

    /** @param {number} minutes */
    minus(minutes) {
        return new Clock(this.#hours, this.#minutes - minutes);
    }

    /** @param {Clock} other */
    equals(other) {
        return this.#hours === other.#hours && this.#minutes === other.#minutes;
    }
}

File: solutions/lua/clock/1/clock.lua
local Clock = {}
Clock.__index = Clock

function normalize(hours, minutes)
    local totalMinutes = hours * 60 + (minutes or 0)
    totalMinutes = ((totalMinutes % (24 * 60)) + 24 * 60) % (24 * 60)
    hours = math.floor(totalMinutes / 60)
    minutes = totalMinutes % 60
    return setmetatable({hours = hours, minutes = minutes}, Clock)
end

function Clock.at(hours, minutes)
    return normalize(hours, minutes)
end

function Clock:__tostring()
    return string.format('%02d:%02d', self.hours, self.minutes)
end

function Clock:plus(minutes)
    return normalize(self.hours, self.minutes + minutes)
end

function Clock:minus(minutes)
    return normalize(self.hours, self.minutes - minutes)
end

function Clock:equals(other)
    return self.hours == other.hours and self.minutes == other.minutes
end

return Clock
